python/pyarrow/tests/test_serialization.py

```diff
@@ -416,3 +416,22 @@ class TempClass(object):
     with pytest.raises(pa.DeserializationCallbackError) as err:
         serialized_object.deserialize(deserialization_context)
     assert err.value.type_id == 20*b"\x00"
+
+
+def test_match_subclasses():
+    # should be able to serialize/deserialize an instance
+    # if a base class has been registered
+    serialization_context = pa.SerializationContext()
+    context.register_type(Foo, "Foo")
+    context.register_type(Bar, "Bar")
+    context.register_type(Baz, "Baz")
+    context.register_type(Qux, "Quz")
+
+    subqux = SubQux()
+    # should fallbact to Qux serializer
+    serialized_object = pa.serialize(subqux, serialization_context)
+
+    reconstructed_object = serialized_object.deserialize(
+        serialization_context
+    )
+    assert type(reconstructed_object) == Qux
```

Comment:
IIUC, without any special handling in [`_deserialize_callback`](https://github.com/dhirschfeld/arrow/blob/1fc2479cdd60493ebf5a8a4a8c556a63d121257f/python/pyarrow/serialization.pxi#L121) you'll lose the information about the particular subclass.

I've punted on this for now because I'm not entirely sure how you could do this robustly in a generic fashion.

In my own code I'm thinking of providing a custom deserializer and storing the subclass information in the serialized object but I'm not sure this is a good *general* solution.